torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,763 @@
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.ACT, [shard_sig]
+                    )
+            placeholder_idx += 1
+        elif node.op == "call_function":
+            # Annotate node types for the computation graph
+            # Data Parallel node propagation logic:
+            # param (non-compute) -> out: param
+            # grad (non-compute before/after) -> out: grad
+            # state -> output: state
+            #
+            # param + activation (param must be replicate, act be sharded) -> out: activation
+            # param/state + grad (param/state/grad be the same spec) -> out: param/state
+            # param + state -> out: param
+
+            if node.target in non_compute_ops:
+                input_full_reduction = False
+                input_nodes = node.all_input_nodes
+                assert (
+                    len(input_nodes) == 1
+                ), f"non-compute op only support one input now, found node: {node} with length of inputs: {len(node.args)}"
+                arg_node_type = dp_strategy_map[input_nodes[0]].node_type
+                if dp_strategy_map[input_nodes[0]].reduction_over_batch:
+                    input_full_reduction = True
+
+                if arg_node_type == NodeType.PARAM:
+                    replica_sig = _gen_replicate_strategy(mesh)
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.PARAM, [replica_sig]
+                    )
+                elif arg_node_type == NodeType.GRAD:
+                    partial_sig = _gen_partial_strategy(mesh)
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.GRAD, [partial_sig]
+                    )
+                elif arg_node_type == NodeType.ACT:
+                    arg_node_spec, _ = batch_dim_analzer.get_batch_dim_shard_spec(
+                        input_nodes[0], mesh
+                    )
+
+                    output_spec, _ = batch_dim_analzer.get_batch_dim_shard_spec(
+                        node, mesh, input_full_reduction
+                    )
+
+                    shard_sig = PlacementStrategy(
+                        output_spec=output_spec, input_specs=[arg_node_spec]
+                    )
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.ACT, [shard_sig]
+                    )
+                else:
+                    raise RuntimeError(
+                        f"non compute op not supporting {arg_node_type}! "
+                    )
+
+                # finished processing this non-compute node
+                continue
+
+            # for computatation nodes, we need to check all the inputs
+            input_args = node.all_input_nodes
+            input_node_types = pytree.tree_map(
+                lambda x: dp_strategy_map[x].node_type, input_args
```

Comment:
nit: `all_input_nodes` already returns a list of nodes, we can save flatten and unflatten from `tree_map` and directly do `input_node_types = [dp_strategy_map[a].node_type for a in input_args]`?